### ISSUE TITLE
Add command line option to change SAVE directory

### ISSUE THREAD

**Issue by shred** [z00m128/sjasmplus#102]
**Is your feature request related to a problem?**
I am invoking SjASMPlus from a build tool. When SjASMPlus is invoked, the current working directory is the directory where all source files are present. The files generated by `SAVE` directives, however, shall be saved to a separate target directory. The target directory should not be hard-coded in the source code, but passed in by a command line option.

**Describe the suggested solution:**
Assume there is a source file `game.asm` that contains this line:
```
SAVESNA "game.sna", START
```

Current situation:
```
sjasmplus game.asm
```
-> "game.sna" is created in the current directory.

Suggested solution:
```
sjasmplus --save=../dist game.asm
```
-> "game.sna" would now be created in `../dist/game.sna`.

This option should be applied to all files generated by all `SAVE` directives.

**Comment by ped7g** [comment]
hmm... sounds sort of useful, may get added.

possible workarounds:
I believe changing to target directory and build from there should work, i.e.
```
cd ../dist; sjasmplus ../src/game.asm
```

Or you can pass target filenames as DEFINE, i.e.
`sjasmplus -D_SAVE_OUT_NAME_=\"../dist/game.sna\" game.asm`
and in source `SAVESNA _SAVE_OUT_NAME_ , START`

Or if you really prefer to pass just target directory path, and build filename in source, you can use Lua script to merge the `-D` define and desired filename in the source, set final filename define in Lua script, and then use that for `SAVE...` directive.

**Comment by shred** [comment]
Thank you!

Choosing an output directory is useful if you want to keep source files and generated files in separate directories, so you can commit the source directory to git, but completely ignore the output directory. It also helps keeping the source directory uncluttered.

After using sjasmplus a little more, I also found other directives like `CSPECTMAP`.

Now I think it would be even more useful to have a kind of `--out` option instead, where you can pass a base directory for *all* files that are generated by the assembler.

Your workaround `cd ../dist; sjasmplus ../src/game.asm` is actually a good solution, but it would look for include files in `dist`, so I'd have to add `../src` to the path of include files, too. I think a line like `sjasmplus --out ../dist game.asm` would be more concise.

**Comment by ped7g** [comment]
Has quite some resistance...

I have some prototype now allowing for `--outprefix=<path>` option, which can be actually any kind of prefix, not just directory, i.e. in your example it would be `sjasmplus --outprefix=../dist/ game.asm` -> needs ending slash to produce `../dist/game.sna`. So far, seems OK-ish to me, whether you find that as feature or bug depends, as capability to add non-dir prefix to the filenames can be useful too.

BUT 1) if the `../dist/` dir does not exist, the opening of file will fail... I was checking if I can add some auto-dir-create functionality to sjasmplus, but the minimum-effort solution is C++17, so I guess I will leave that responsibility to the build script (I don't want to reinvent `mkdir -p` myself).

BUT 2) it affects all output filenames defined in the source, but not the ones provided on command line, like `--raw=game.sna` or `--lst=game.lst`, so these are again responsibility of build script.

BUT 3) auto-names for `-lst` and `--sld` ignore the prefix too.

In this form the feature feels to me a bit unfinished and sloppy, but I don't see any obvious way to improve it, as there's so many "what if" possible scenarios (what if you specify full path in the save directive like `savebin "/tmp/table.bin"`? It will become prefixed... and this makes at least some sense with NIX-like paths, but in windows way you will end with path `../dist/n:/tmp/table.bin` ... then again, I refuse to worry about windows users, as it's their choice to use weird OS, so I guess they will have to figure out these corner cases themselves).

So I guess I will add it in the prototype form, as I described it above.

**Comment by shred** [comment]
Thank you for adding the feature!

1) I also wouldn't expect the assembler to create the target directory. The build tool should take care of that if necessary.
2) That's okay, as these paths could easily be changed in the command line. This way, it won't even break existing build scripts if the `--outprefix` option is added later.

**Comment by oxidaan** [comment]
--outprefix=<path> doesn't seem to work when <path> contains a space. Putting it in quotes like --outprefix="c:\path\to\dir with spaces" not working either...

**Comment by ped7g** [comment]
@oxidaan can't reproduce here, works for me.

I extended the `--outprefix` test to use space in path, and it works in all CI builds (including windows). I use backslash to escape it like this:
`--nologo --msg=war --fullpath --outprefix=out\ dir/ ...`   (this is inside bash script).

I tried also quotes `--outprefix="out dir/"` in interactive bash terminal on linux, that did work too, although I was unable to figure out correct combination of escaping of quotes inside the auto-test bash script :man_shrugging: 

I don't know how to escape the space correctly in windows cmd.exe or bat files, but as long as sjasmplus receives the correct full string with path in arguments, I can't imagine reason why it would fail in the file-name-building code.

One last idea, did you put the trailing slash after the directory name? Ie.
`sjasmplus --outprefix="c:/path with spaces/" x.asm`
That should produce filenames like "c:/path with spaces/x.sna", while
`sjasmplus --outprefix="c:/path with spaces" x.asm`
will produce filenames like "c:/path with spacesx.sna" -> prefixing the file name with the string.

This is not a bug, it's a feature, the outprefix is literally that, prefix, can also extend filenames themselves, so folder-only prefix has to end with slash.
